### ISSUE TITLE
improve timing printing in python

### ISSUE THREAD

**Issue by AnderBiguri** [CERN/TIGRE#284]
reduce decimals, and dont just print (s) but a human readable time in H:M:S:ms

**Comment by Ayushsunny** [comment]
Hello Sir @AnderBiguri, can you guide me on this? I would like to work on this issue

**Comment by AnderBiguri** [comment]
Hi @Ayushsunny , absolutely! 

This one is quite easy:

Here: 

https://github.com/CERN/TIGRE/blob/master/Python/tigre/algorithms/art_family_algorithms.py#L103

and maybe a couple of other places, we use `toc-tic` to print the time, in seconds. Ideally, we just want some function that takes the time in seconds, and converts it to HH:mm:ss:ms  kind of format, so its easier to read. We want it to print something like:

"Esitmated time until completetion :  XXh XXm XXs XXXXms. "

Perhaps we can simply just remove h,m,s whenever they are 0.

**Comment by Ayushsunny** [comment]
So I need to make this function right ?

```python
import time
time.strftime('%H:%M:%S', time.gmtime(12345))
'03:25:45'
```
but I didn't understand that 
`The best way is probably to make a new .py file within Utilities folder, that takes time in s and returns a string.
MATLAB has already a function (you can test it in MATLAB or Octave if you don;t have a MATLAB function):`

**Comment by AnderBiguri** [comment]
@Ayushsunny ah, if there is an inbuilt function for it, then that is good enough!
In MATLAB there is not, so I was suggesting writting it up, but if `time.strftime` is what we want, the lets just use that! no need to make a new .py file :)

**Comment by Ayushsunny** [comment]
> @Ayushsunny ah, if there is an inbuilt function for it, then that is good enough!
> In MATLAB there is not, so I was suggesting writting it up, but if `time.strftime` is what we want, the lets just use that! no need to make a new .py file :)

so this is the tic-toc function https://github.com/CERN/TIGRE/blob/bde3d952bbc63586928dd5f29e4fa1be30aed573/Python/tigre/algorithms/art_family_algorithms.py#L83-L104

we are going to change sir @AnderBiguri ?

**Comment by AnderBiguri** [comment]
Specifically this part:

```
print( 
                         "Esitmated time until completetion (s): " 
                         + str((self.niter - 1) * (tic - toc)) 
                     ) 
```

I suggest running demo 4 of the python code, that will use this printing, and you will see it prints in `s` but we want a more human readable format.

**Comment by AnderBiguri** [comment]
We also want the same thing at:

https://github.com/CERN/TIGRE/blob/bde3d952bbc63586928dd5f29e4fa1be30aed573/Python/tigre/algorithms/ista_algorithms.py#L178-L181

and 

https://github.com/CERN/TIGRE/blob/bde3d952bbc63586928dd5f29e4fa1be30aed573/Python/tigre/algorithms/iterative_recon_alg.py#L324-L327

and

https://github.com/CERN/TIGRE/blob/bde3d952bbc63586928dd5f29e4fa1be30aed573/Python/tigre/algorithms/krylov_subspace_algorithms.py#L70-L73

https://github.com/CERN/TIGRE/blob/bde3d952bbc63586928dd5f29e4fa1be30aed573/Python/tigre/algorithms/pocs_algorithms.py#L147-L150

and finally:

https://github.com/CERN/TIGRE/blob/bde3d952bbc63586928dd5f29e4fa1be30aed573/Python/tigre/algorithms/statistical_algorithms.py#L51-L55

I think that is all of them, but if you find any other, change it too :)

**Comment by Ayushsunny** [comment]
Okay sir I got it. So I need to replace this str -
```py
print( 
     "Esitmated time until completetion (s): " 
     + str((self.niter - 1) * (tic - toc)) 
 ) 
```

with this -
```py
print( 
     "Esitmated time until completetion (HH:mm:ss): " 
     + time.strftime('%H:%M:%S', time.gmtime(tic - toc))) 
 ) 
```
if I am not wrong ?
please clarify @AnderBiguri

**Comment by AnderBiguri** [comment]
Yes, I suppose! You need to test it to see if it works, of course.

**Comment by AnderBiguri** [comment]
Before you can contribute to the repository you need to be able to run the code in you machine. Follow the installation instructions please.

**Comment by Ayushsunny** [comment]
> Before you can contribute to the repository you need to be able to run the code in you machine. Follow the installation instructions please.
> […](#)
> On 25 Apr 2021, 21:30 +0200, Ayush Anand ***@***.***>, wrote: > Yes, I suppose! You need to test it to see if it works, of course. > How to run these specific on my local system, sir? > I am using Ubuntu 20.04 > —You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

Sir CUDA is so big file for my local system.Can you please check it on yours?
please sir @AnderBiguri

**Comment by AnderBiguri** [comment]
Unfortunately no...
If you want to help fixing code in Tigre then you will need to run Tigre on your local system, add your code, test it, and make a pull request when it works.

Otherwise I would just fix it myself!

**Comment by Ayushsunny** [comment]
> Unfortunately no... If you want to help fixing code in Tigre then you will need to run Tigre on your local system, add your code, test it, and make a pull request when it works. Otherwise I would just fix it myself!

Sorry Sir, but I really wanted to be a long term contributor in CERN.
**okay sir fix it yourself, sorry again.**

**Comment by Ayushsunny** [comment]
> Sorry, you seem to misunderstand @Ayushsunny ....
> 
> I am very happy for you to be a long term contributor in TIGRE.
> However, being a contributor literally means that you can run the code, and you will submit tested code via PRs.
> I am happy for you to do this, and this is why I used so much of my time to explain to you where to contribute, rather than doing it myself, because I am happy if you do it and become a contributor.
> 
> > Can you please check it on yours?
> 
> But you asked me to check it myself, and that is just doing it myself, that would not make you a contributor! to do so you need to submit the code after testing it yourself, that is the definition.
> 
> All the issues in TIGRE are solvable, and I can solve them, I just don't have the time to work on them. So I am very happy to accept contributors because I want this to be an open source software for everyone to use and contribute to!
> 
> So up to you really, but its impossible to contribute to an open source project without being able to run the code!
> 
> Note: this is not CERN, I do not work for CERN, nor TIGRE is officially _maintained_ by CERN, they currently just host it and hold the license.

Thank you So much, Sir for your Positive response.
I'll do something for CUDA and anyhow I'll test my code and will reply you here .
again thank you so much @AnderBiguri

**Comment by lawful02** [comment]
Sir @AnderBiguri , is this issue still open? If so, I would like to work on this issue.

**Comment by AnderBiguri** [comment]
@lawful02  not sure if @Ayushsunny is working on it or not, I haven't heard anything that is not written here. 

If he does not reply then sure, work on it. There are also some other good first issues that are open to tackle, if you prefer. Its the same to me (in general, you don't need my permission to do anything ;) just make a PR, I will review it and merge it if good).

**Comment by lawful02** [comment]
Alright sir, in that case i will wait for @Ayushsunny to reply, while i look into other issues.

**Comment by Ayushsunny** [comment]
Hello @AnderBiguri Sir,
why you have closed this [PR](https://github.com/CERN/TIGRE/pull/290#event-4690533122) without any review or request changes.?

**Comment by tsadakane** [comment]
@AnderBiguri 
> a human readable time in H:M:S:ms

ISO 8601 ("https://www.iso.org/obp/ui/#iso:std:iso:8601:-1:ed-1:v1:en" 3.2.6   Separator symbols )says

“:” (colon) | the “:” colon character, in extended format, separates the time scale components for “hour” and “minute”, and “minute” and “second”.
-- | --
“.” (period), “,” (comma) | the “.” period and “,” comma characters are decimal sign used to separate the integer part from the decimal fraction of a number.

I would like to suggest to use "H:M:S.ms". What do you think?

**Comment by AnderBiguri** [comment]
I have my reasons to close it. 
Please  @Ayushsunny , make a new PR, with @tsadakane 's suggestions. Also make the text
` "Estimated time until completion:" ` with the new time format.

**Comment by AnderBiguri** [comment]
@tsadakane Thanks for the suggestion again.

I don't understand well the behavior of these users, its been a bit frustrating. 
This issue is currently unsigned, if you want to give it a shot go for it (as you have suggested a very good format :) ).
Otherwise I will do it myself over the weekend, just to close this issue once for all.

**Comment by tsadakane** [comment]
@AnderBiguri Thank you for your proposal, but unfortunately I am not going to work on this issue. If you allow me to do another suggestion, I think the precision of the prediction is not so high as ms, so H:M:S is enough.

**Comment by AnderBiguri** [comment]
@tsadakane  agreed! In MATLAB I don't have ms either. I will do it thanks :+1:
